controllers/edgedeployment_controller.go

```diff
@@ -64,71 +66,160 @@ func (r *EdgeDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Reque
-	blockOwnerDeletion := true
-	isController := false
-	return metav1.OwnerReference{
-		APIVersion:         ed.GroupVersionKind().GroupVersion().String(),
-		Kind:               ed.GetObjectKind().GroupVersionKind().Kind,
-		Name:               ed.GetName(),
-		UID:                ed.GetUID(),
-		BlockOwnerDeletion: &blockOwnerDeletion,
-		Controller:         &isController,
+func (r *EdgeDeploymentReconciler) removeDeploymentFromDevice(ctx context.Context, name string, edgeDevice managementv1alpha1.EdgeDevice) error {
+	var newDeployments []managementv1alpha1.Deployment
+	for _, deployment := range edgeDevice.Status.Deployments {
+		if deployment.Name != name {
+			newDeployments = append(newDeployments, deployment)
+		}
+	}
+	edgeDevice.Status.Deployments = newDeployments
+	err := r.EdgeDeviceRepository.UpdateStatus(ctx, &edgeDevice)
+	if err != nil {
+		return err
+	}
+
+	deviceCopy := edgeDevice.DeepCopy()
+	if deviceCopy.Labels != nil {
+		delete(deviceCopy.Labels, name)
+		err = r.EdgeDeviceRepository.Patch(ctx, &edgeDevice, deviceCopy)
+		if err != nil {
+			return err
+		}
 	}
+	return nil
 }
 
-func (r *EdgeDeploymentReconciler) addOwnerReference(ctx context.Context, owner *managementv1alpha1.EdgeDevice, ownedDeployment *managementv1alpha1.EdgeDeployment) error {
-	ownerRefs := ownedDeployment.GetOwnerReferences()
-	if ownerRefs == nil {
-		ownerRefs = []metav1.OwnerReference{}
-	}
-	if hasOwnerReference(ownerRefs, owner) {
-		return nil
+func (r *EdgeDeploymentReconciler) removeDeploymentFromNonMatchingDevices(ctx context.Context, name string, matchingDevices, labelledDevices []managementv1alpha1.EdgeDevice) error {
+	matchingDevicesMap := make(map[string]struct{})
+	for _, device := range matchingDevices {
+		matchingDevicesMap[device.Name] = struct{}{}
 	}
-	ownerReference := newEdgeDeviceOwnerReference(owner)
 
-	ownerRefs = append(ownerRefs, ownerReference)
+	for _, device := range labelledDevices {
+		if _, ok := matchingDevicesMap[device.Name]; !ok {
+			err := r.removeDeploymentFromDevice(ctx, name, device)
+			if err != nil {
+				return err
+			}
+		}
+	}
 
-	newEdgeDeployment := ownedDeployment.DeepCopy()
-	newEdgeDeployment.SetOwnerReferences(ownerRefs)
+	return nil
+}
 
-	return r.EdgeDeploymentRepository.Patch(ctx, ownedDeployment, newEdgeDeployment)
+func (r *EdgeDeploymentReconciler) addDeploymentsToDevices(ctx context.Context, name string, edgeDevices []managementv1alpha1.EdgeDevice) error {
+outer:
+	for _, edgeDevice := range edgeDevices {
+		for _, deployment := range edgeDevice.Status.Deployments {
+			if deployment.Name == name {
+				// Move to the next device
+				continue outer
+			}
+		}
+		deploymentStatus := managementv1alpha1.Deployment{Name: name, Phase: managementv1alpha1.Deploying}
+		edgeDevice.Status.Deployments = append(edgeDevice.Status.Deployments, deploymentStatus)
+		err := r.EdgeDeviceRepository.UpdateStatus(ctx, &edgeDevice)
```

Comment:
I think that `EdgeDeviceRepository.UpdateStatus` needs to fetch the device status first and then apply the change on it to minimize chances of failures (e.g. if 2 deployments were updated, or if the status is updated as a result of a heartbeat).

Else, we might face "cannot update resource, not latest version" kind of errors

Reply:
`EdgeDeviceRepository.UpdateStatus` can't just copy desired status to object freshly retrieved from etcd because the object coming in the parameter may have something in the status sub-resource that is gone in the latest repository version and we could restore it accidentally (especially `status.deployments` will be problematic). 
There are two options here: 
 - let the caller resolve conflict and then retry updating status
 - let's return error and make the reconciliation re-run

I went with the second option as it feel as a more natural way of dealing with problems in a controller.